### DISCUSSION TITLE
How to attach request/response interceptors to fetch

### DISCUSSION THREAD

**Discussion by AndonMitev** [vercel/next.js#57738] [Help] [answered]
### Summary

I have looked for some workaround but not able to find something that works properly, maybe closest thing to this was `monkey-patch`, so i'm asking here if someone else have managed to attach request or response interceptor to the native fetch can please share more information how he has done it?

**Comment by AndonMitev** [comment]
So far i have created something like this:

```
export const fetchApi = async (url: RequestInfo | URL, data?: any) => {
  const session = await getSession();

  const init = data
    ? {
        body: JSON.stringify(data),
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${session?.user.accessToken}`
        }
      }
    : undefined;

  const response = await fetch(url, { ...init });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error('Unauthorized');
    }

    throw new Error('Request Failed');
  }

  return response.json();
};
```

but i'm not sure if this is the right direction

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I built for a hackathon, a simple UI that could should us Server Side, the requests that happen in React Server Components setups.

The key insight for the hack was the existence of this package (which is a building block of MSW itself):

> Note that __I do not own this package__! ❗❗

- https://www.npmjs.com/package/@mswjs/interceptors

```shell
npm install @mswjs/interceptors
```
Adjust the command to yarn or pnpm.

And you'd use it like:

```tsx
// app/interceptors.ts
import { BatchInterceptor } from "@mswjs/interceptors";
import { ClientRequestInterceptor } from "@mswjs/interceptors/ClientRequest";
import { XMLHttpRequestInterceptor } from "@mswjs/interceptors/XMLHttpRequest";
import { FetchInterceptor } from "@mswjs/interceptors/fetch";

const interceptor = new BatchInterceptor({
  name: "my-interceptor",
  interceptors: [
    new ClientRequestInterceptor(),
    new XMLHttpRequestInterceptor(),
    new FetchInterceptor(),
  ],
});

interceptor.apply();

// This "request" listener will be called on both
// "http.ClientRequest" and "XMLHttpRequest" being dispatched.
interceptor.on("request", console.log);
interceptor.on("response", console.log);
```

And then just import that into your layout:

```tsx
// app/layout.tsx
import "./interceptors";
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

// rest of layout file

```

Note that I kept this for a hackathon and under dev mode only :) you might have to do tweaks of your own. The interceptor you also really care about is: `FetchInterceptor`

**Comment by AndonMitev** [comment]
Hey @icyJoseph thanks for the answering. BTW Amazing job wdym is under dev mode only there are 2.2mil weekly downloads? :D 

is it FetchInterceptor applicable for server and clients components, will try it asap. My goal is to:

1. Attach Authorization header to the request
2. Handle 401 or such generic errors on the clients, is FetchInterceptor the right package for me?

Please say yes 😂 . BTW once again amazing job with your package bro

**Reply by icyJoseph** [reply]
Oh no, that's not my package! I used it to build a hackathon project!

**Reply by icyJoseph** [reply]
And yeah it does support browser environments.

It even has a preset for it:

- https://www.npmjs.com/package/@mswjs/interceptors#browser-preset

**Reply by AndonMitev** [reply]
ahh i see, thank you for the information and finally which one you suggest to be used `mswjs` vs `fetch-interceptor` have you used both in next@13+?

**Reply by icyJoseph** [reply]
I'd go with `msw` packages, they keep on developing and improving their API's, deliver fixes and such. 

The original MSW uses service workers to intercept requests in tests, dev, prod, etc, environments, you define it, and it leads to for example, testing without mocking.

Hope it does the trick for you ~ and please don't forget to mark as answered, or ask further.

**Reply by AndonMitev** [reply]
I wont, ty so much will try it! Again thank you so much!
